cmd/clusterctl/pkg/client/cluster/upgrader.go

```diff
@@ -0,0 +1,275 @@
+	var availableVersions []version.Version
+	for _, availableVersion := range allAvailableVersions {
+		// we are ignoring the conversion error here because a first check already passed above
+		availableSemVersion, _ := version.ParseSemantic(availableVersion)
+
+		if currentSemVersion.LessThan(availableSemVersion) {
+			if isInReleaseSeries(availableSemVersion, providerMetadata) {
+				availableVersions = append(availableVersions, *availableSemVersion)
+				continue
+			}
+			return nil, nil, errors.Errorf("invalid provider metadata: version %s (one of the available versions) for the provider %s/%s does not match any release series", availableVersion, provider.Namespace, provider.Name)
+		}
+	}
+
+	sort.Slice(availableVersions, func(i, j int) bool {
+		return availableVersions[i].LessThan(&availableVersions[j])
+	})
+
+	return providerMetadata, availableVersions, nil
+}
+
+// getClusterAPIVersionForUpgrade return the list of ClusterAPI version available for a provider upgrade. e.g.
+// - the current version of the provider is linked with v1alpha3 ClusterAPIVersion (the latest) --> v1alpha3
+// - the current version of the provider is linked with v1alpha3 ClusterAPIVersion but there is also v1alpha4 ClusterAPIVersion defined --> v1alpha3, v1alpha4
+func getClusterAPIVersionForUpgrade(currentVersion version.Version, metadata *clusterctlv1.Metadata) []string {
+	// Sorts release series; this ensures also an implicit ordering of ClusterAPI version.
+	sort.Slice(metadata.ReleaseSeries, func(i, j int) bool {
+		return metadata.ReleaseSeries[i].Major < metadata.ReleaseSeries[j].Major ||
+			(metadata.ReleaseSeries[i].Major == metadata.ReleaseSeries[j].Major && metadata.ReleaseSeries[i].Minor < metadata.ReleaseSeries[j].Minor)
+	})
+
+	var clusterAPIVersionsForUpgrade []string
+	clusterAPIVersions := make(map[string]interface{})
+	for _, releaseSeries := range metadata.ReleaseSeries {
+		// Drop the release series if older that the current one
+		if currentVersion.Major() > releaseSeries.Major || ((currentVersion.Major() == releaseSeries.Major) && (currentVersion.Minor() > releaseSeries.Minor)) {
+			continue
+		}
+
+		// Stores the Cluster APIVersion, without duplicates.
+		if _, value := clusterAPIVersions[releaseSeries.ClusterAPIVersion]; !value {
+			clusterAPIVersions[releaseSeries.ClusterAPIVersion] = nil
+			clusterAPIVersionsForUpgrade = append(clusterAPIVersionsForUpgrade, releaseSeries.ClusterAPIVersion)
+		}
+	}
+
+	return clusterAPIVersionsForUpgrade
+}
+
+// getUpgradeVersion returns the target upgrade version for a provider/targetClusterAPIVersion.
+// the target upgrade version is tha latest version available in the list of available versions for the target Cluster APIVersion.
+func getUpgradeVersion(currentVersion version.Version, targetClusterAPIVersion string, availableVersions []version.Version, metadata *clusterctlv1.Metadata) *version.Version {
```

Comment:
```suggestion
func getUpgradeVersion(currentVersion version.Version, targetClusterAPIVersion string, availableVersions []*version.Version, metadata *clusterctlv1.Metadata) *version.Version {
```

Reply:
nit

Reply:
After the refactor this does not apply anymore